test/profiler/test_profiler.py

```diff
@@ -1370,6 +1371,265 @@ def get_fields(op_name, index):
         self.assertEqual(c_id, c_id_new)
         self.assertEqual(d_id, c_id_new)
 
+    @staticmethod
+    def _format_allocations(prof: torch.profiler.profile):
+        root_events = prof.profiler.kineto_results.experimental_event_tree()
+        events = sorted(_utils.traverse_dfs(root_events), key=lambda x: x.start_time_ns)
+        allocations = tuple(
+            event.extra_fields
+            for event in events
+            if isinstance(event.extra_fields, torch._C._profiler._ExtraFields_Allocation)
+        )
+
+        return textwrap.indent("\n".join(
+            f"{repr(i.id):>5}{' ' * 6}"
+            f"{'Allocation' if i.alloc_size > 0 else 'Free'}"
+            for i in allocations
+        ), " " * 12)
+
+    def test_tensorimpl_invalidation_set(self) -> None:
+        with profile(profile_memory=True, record_shapes=True) as p:
+            x = torch.ones((1,))
+            x.set_(torch.ones((1,)).storage())
+            x.view_as(x)
+
+            del x
+            gc.collect()
+
+        self.assertExpectedInline(
+            self._format_allocations(p),
+            """\
+                0      Allocation
+                0      Allocation
+                0      Free
+                0      Free"""
+        )
+
+    def test_tensorimpl_invalidation_set_breaking(self) -> None:
+        with profile(profile_memory=True, record_shapes=True) as p:
+            x = torch.ones((1,))
+
+            # This call breaks the chain by freeing the old storage before the
+            # new one is allocated. (And so they should not share an ID.)
+            x.set_()
+
+            x.set_(torch.ones((1,)).storage())
+            x.view_as(x)
+
+            del x
+            gc.collect()
+
+        self.assertExpectedInline(
+            self._format_allocations(p),
+            """\
+                0      Allocation
+                0      Free
+                1      Allocation
+                1      Free"""
+        )
+
+    def test_tensorimpl_invalidation_keep_alive(self) -> None:
+        with profile(profile_memory=True, record_shapes=True) as p:
+            x = torch.ones((1,))
+            x_storages = [x.storage()]
+            for _ in range(3):
+                x.set_()
+                x.set_(torch.ones((1,)).storage())
+
+                # This keeps the StorageImpls alive and preserves the chain.
+                # (Despite the `set_()` call.)
+                x_storages.append(x.storage())
+            x.view_as(x)
+
+            # Free storage in a deterministic fashion.
+            while x_storages:
+                x_storages.pop()
+                gc.collect()
+
+            for _ in range(3):
+                x.set_(torch.ones((1,)).storage())
+            x.view_as(x)
```

Comment:
Profiler only sees a Tensor (and therefore the `(TensorImpl, StorageImpl)` pair) when a Tensor is used in an op. Allocation just sees `void*`. So you have to do something with the Tensor to actually plumb it through an op, otherwise profiler will never see the new storage of `x` after the last `set_`. `x.view_as(x)` is basically a no-op. (I'll put that behind a staticmethod and write a good docstring.)